presto-parquet/src/main/java/com/facebook/presto/parquet/reader/BinaryColumnReader.java

```diff
@@ -52,6 +55,9 @@ protected void readValue(BlockBuilder blockBuilder, Type type)
             if (isCharType(type)) {
                 value = truncateToLengthAndTrimSpaces(value, type);
             }
+            if (type instanceof UuidType) {
+                value = wrappedLongArray(rawUuidValuesFromBigEndian(binary.getBytes()));
```

Comment:
There are a few opportunities to improve the perf and reliability here:
1. The memory for `value` was pre-allocated above on line 50 `value = wrappedBuffer(binary.getBytes());`. In this line `binary.getBytes()` allocates the memory and copy the data once already. But your `rawUuidValuesFromBigEndian` was not making use of the memory in `binary` object, but instead, it re-allocates two longs inside. Note that for char and varchar types, the memory was re-used and no new allocation was done. For UUID type, the output value is always a fixed length of 2 longs.  This is different than char and varchar types because the output length can only be known at run time for these two types. So for UUID, we can at least avoid the double allocation. I would suggest handle UUID separately and do not allocate the `Slice value` above.
2. It is not good and not necessary to wrap the 2 longs as a Slice at all. `wrappedLongArray` allocates a new Slice object. Although the base array was not copied, new memory needs to be allocate to hold the Slice object itself. Then `type.writeSlice(blockBuilder, value);` peels the Slice just to get the 2 longs we just wrapped up. I've seen many cases where allocating so many Slices objects caused GC pressure and perf degradation. It would be better to just use `writeLong` twice here. I know this is a pre-existing design problem, and would recommend fix this in a separate commit.

Reply:
Thanks for your suggestions. I tried to reduce the allocations and copying by writing directly to the block builder as I pulled values from the buffer. It increased the performance a measurable amount for the non-batchreader implementations

old perf results
```
Benchmark                       (enableOptimizedReader)  (parquetEncoding)   Mode  Cnt    Score   Error  Units
BenchmarkUuidColumnReader.read                     true              PLAIN  thrpt   10  158.193 ± 1.058  ops/s
BenchmarkUuidColumnReader.read                     true   DELTA_BYTE_ARRAY  thrpt   10   32.211 ± 0.345  ops/s
BenchmarkUuidColumnReader.read                    false              PLAIN  thrpt   10   13.030 ± 0.463  ops/s
BenchmarkUuidColumnReader.read                    false   DELTA_BYTE_ARRAY  thrpt   10   10.121 ± 0.377  ops/s
```

new perf results
```
Benchmark                       (enableOptimizedReader)  (parquetEncoding)   Mode  Cnt    Score   Error  Units
BenchmarkUuidColumnReader.read                     true              PLAIN  thrpt   20  161.496 ± 1.250  ops/s
BenchmarkUuidColumnReader.read                     true   DELTA_BYTE_ARRAY  thrpt   20   33.104 ± 0.257  ops/s
BenchmarkUuidColumnReader.read                    false              PLAIN  thrpt   20   27.478 ± 1.132  ops/s
BenchmarkUuidColumnReader.read                    false   DELTA_BYTE_ARRAY  thrpt   20   20.779 ± 0.564  ops/s
```

Reply:
The updated code is here:

- [BinaryColumnReader](https://github.com/prestodb/presto/blob/891af15747ea2f43f9f2e6a65968a0c5ea6e73d3/presto-parquet/src/main/java/com/facebook/presto/parquet/reader/BinaryColumnReader.java#L46-L48)
- [Uuids](https://github.com/prestodb/presto/blob/891af15747ea2f43f9f2e6a65968a0c5ea6e73d3/presto-common/src/main/java/com/facebook/presto/common/type/Uuids.java#L76-L92)